core/tee/tee_svc_storage.c

```diff
@@ -542,6 +550,12 @@ TEE_Result syscall_storage_obj_del(unsigned long obj)
 	if (o->pobj == NULL || o->pobj->obj_id == NULL)
 		return TEE_ERROR_BAD_STATE;
 
+#if defined(CFG_NXP_SE05X_SVC)
+	len = o->info.dataSize > sizeof(data) ? sizeof(data) : len;
+	res = o->pobj->fops->read(o->fh, o->info.dataPosition, data, &len);
+	if (res == TEE_SUCCESS)
+		se050_delete_persistent_key(data, len);
+#endif
```

Comment:
IIUC, the call syscall_storage_obj_del is mapped to TEE_CloseAndDeletePersistentObject1.  Was the intention to allow creation of persistent object in SE using the TEE_CreatePersistantObject ? There has been no creation of an object in SE with this call so if I have understood this correctly, I am not sure if this would work.

Reply:
not sure what you mean but yes this is tested with ECC and it does work (we use it heavily with pkcs11)

The object would have been created already in the SE persistent storage using the cipher api (https://github.com/ldts/optee_os/blob/se050-plug_and_trust/lib/libnxpse050/core/ecc.c#L418).

Reply:
I am not sure I agree about it being a better approach....there is a fundamental difference in the fact that this driver is available to all platforms where as  CAAM is only available to some iMX platforms. In relation to this I am still wondering where am I going to set all the configs needed (will I have to modify all platform-*/conf.mk to integrate this?...it seems a bit of an overkill)

So really, this is a crypto service in exact the same way than mbed or tomcrypt. the only issue really is that mbed and tomcrypt havent been intregrated in a way that allows littel flexibility for a third library to join the party without a big rework...

Reply:
> yes this is tested with ECC and it does work (we use it heavily with pkcs11)
> 
> I need to trace the logic to key creation but ultimately it reaches the cipher api where the driver is instructed to create the key. The issue I had was that the object deletion never reaches cipher hence the problem to delete and this particular hack.
> 
> The object would have been created already in the SE persistent storage using the cipher api (https://github.com/ldts/optee_os/blob/se050-plug_and_trust/lib/libnxpse050/core/ecc.c#L418).

Thanks for the link.
What I meant was that this change in is in a function which gets called from TA's which makes a call to TEE_CloseAndDeletePersistentObject1. When the SE Config is enabled by a platform, there may be TA's creating and destroying Persistent objects for different usecases not linked with ECC key storage in SE. I may be missing something here but just wanted to be sure that this change won't impact such use-cases ?

Reply:
no no, this is just a hack (protected by the SE050 config) to highlight an issue, it cant be merged like this...:the issue is the apparent need for the TEE_CloseAndDeletePersistentObject API to be able to interface to the crypto API in case the object is being stored by crypto.
This is relatively safe: basically the codes grabs the data, and calls the se050 adaptor so it can search for persistent key ids in the buffer (they are watermarked)...if there are any, it gets the key ids and calls the SE050 for deletion (otherwise the kesy persist in the SE050 NVM)

Reply:
@jforissier @jenswi-linaro does this need further clarification? the patch intentionally dirty to highlight the bigger question. 

A more generic implementation could be something like this where crypto_storage_obj_del would be a new __weak crypto interface

```
TEE_Result syscall_storage_obj_del(unsigned long obj)
{
	TEE_Result res;
	struct tee_ta_session *sess;
	struct tee_obj *o;
	struct user_ta_ctx *utc;
	uint8_t *data;
	size_t len;

	res = tee_ta_get_current_session(&sess);
	if (res != TEE_SUCCESS)
		return res;
	utc = to_user_ta_ctx(sess->ctx);

	res = tee_obj_get(utc, uref_to_vaddr(obj), &o);
	if (res != TEE_SUCCESS)
		return res;

	if (!(o->info.handleFlags & TEE_DATA_FLAG_ACCESS_WRITE_META))
		return TEE_ERROR_ACCESS_CONFLICT;

	if (o->pobj == NULL || o->pobj->obj_id == NULL)
		return TEE_ERROR_BAD_STATE;

	len = o->info.dataSize;
	data = calloc(1, len);
	if (!data)
		return TEE_ERROR_OUT_OF_MEMORY;

	res = o->pobj->fops->read(o->fh, o->info.dataPosition, data, &len);
	if (res == TEE_SUCCESS)
		crypto_storage_obj_del(data, len);

	free(data)

	res = o->pobj->fops->remove(o->pobj);
	tee_obj_close(utc, o);

	return res;
}
```

Reply:
> @jforissier @jenswi-linaro does this need further clarification?

It does, at least for me ;-) Which TA call is causing the creation of a persistent key in the SE?

Reply:
@jforissier 

ok, I'll tidy it up and repost then. @jenswi-linaro suggested that we could set some flags on the object to decide whether to then call crypto or not.

We use this version of pkcs11:
 https://github.com/foundriesio/optee-sks 
The plan is to use the upstreamed version but AFAIK still lacks some of the functionality that we need.

An example for EC key creation: 
https://github.com/foundriesio/optee-sks/blob/master/ta/secure_key_services/src/processing.c#L606 
calling into
https://github.com/foundriesio/optee-sks/blob/master/ta/secure_key_services/src/processing_ec.c#L1237